[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126678

[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

etaf committed:
[Inductor][Intel GPU] Suppot codegen empty_strided_xpu, align with #118255.

[ghstack-poisoned]

etaf commented:
This PR depends on the PR: https://github.com/intel/torch-xpu-ops/pull/279

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

EikanWang commented:
@etaf , is this PR ready for review?

etaf commented:
> @etaf , is this PR ready for review?

yes, but it depends on the update of torch_xpu_ops, let me make it a draft first.

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

fengyuan14 commented on caffe2/CMakeLists.txt:
```diff
@@ -1076,6 +1076,14 @@ if(USE_XPU)
+  # Set exported torch-xpu-ops header to torch
+  set(INCLUDE_DIR_FOR_TORCH "${CMAKE_BINARY_DIR}/aten/src/ATen/xpu" )
+  file(MAKE_DIRECTORY ${INCLUDE_DIR_FOR_TORCH})
+  file(GLOB xpu_h "${TORCH_ROOT}/third_party/torch-xpu-ops/src/aten/*.h")
```
We should not use hard-coded dir, but use a cmake variable outputed by `add_directory(torch-xpu-ops)`. Since the code will couple with torch-xpu-ops implementation. The variable could align with third_party packages, like MKLDNN_INCLUDE_DIR.

fengyuan14 commented on caffe2/CMakeLists.txt:
```diff
@@ -1076,6 +1076,14 @@ if(USE_XPU)
+  file(MAKE_DIRECTORY ${INCLUDE_DIR_FOR_TORCH})
+  file(GLOB xpu_h "${TORCH_ROOT}/third_party/torch-xpu-ops/src/aten/*.h")
+  foreach(HEADER  ${xpu_h})
+    file(COPY ${HEADER} DESTINATION ${INCLUDE_DIR_FOR_TORCH})
```
You might not copy them but append TORCH_XPU_OPS_INCLUDE_DIR into caffe2/ATen include dirs. Please check cmake of dynamo should depends on caffe2/ATen include dirs. Then sources in dynamo could find the header by nature without copy.

etaf replied:
I designed it this ‘copy' way because I wanted the include path to be consistent with cuda/cpu (`<ATen/xpu/EmptyTensor.h>`) so that it would be more readable.

etaf commented:
@pytorchbot rebase

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

fengyuan14 commented on caffe2/CMakeLists.txt:
```diff
@@ -1068,6 +1068,10 @@ if(USE_XPU)
     endif()
+
+    # Set cached ${ATen_XPU_INCLUDE_DIRS} to torch
+    include_directories(SYSTEM ${ATen_XPU_INCLUDE_DIRS})
```
How do cmake variables of CPU, CUDA or ATen include directories do in PyTorch? Do we align with them?

etaf replied:
They use the the the similar way: https://github.com/pytorch/pytorch/blob/ab9791c0e342753013181eeeab300a05774fc456/CMakeLists.txt#L1205-L1206

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf commented on test/inductor/test_cuda_repro.py:
```diff
@@ -1276,7 +1276,7 @@ def fn(x, y):
 
         # Expect the code iterates in contiguous order, and is not tiled
-        kernel_code = "\n".join(code[0].split("\n")[50:64])
+        kernel_code = "\n".join(code[0].split("\n")[51:65])
```
This PR add a line in the header of generated ouput code: 
```
                empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
                empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
+              empty_strided_xpu = torch._C._dynamo.guards._empty_strided_xpu
```
And this test case need to check the content of ouput code between line 50~64.

EikanWang replied:
Thanks for the elaboration.

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255."


[Inductor][Intel GPU] Support codegen empty_strided_xpu, align with #118255.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf commented:
@pytorchbot merge

etaf commented:
@pytorchbot merge
